File: laskarit/viikko2/Unicafe/src/test/java/com/mycompany/unicafe/KassapaateTest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.mycompany.unicafe;

import static org.hamcrest.core.Is.is;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import static org.junit.Assert.*;

/**
 *
 * @author tuukkala
 */
public class KassapaateTest {

    Kassapaate kassa;
    
    @Before
    public void setUp(){
        this.kassa = new Kassapaate();
    }
    
//    luodun kassapäätteen rahamäärä ja myytyjen lounaiden määrä on oikea (rahaa 1000, lounaita myyty 0)
    @Test
    public void kassanRahamaaraAlussaOikea() {
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
    
    @Test
    public void kassanMyytyjenEdullistenLounaidenMaaraAlussaOikea() {
        assertThat(kassa.edullisiaLounaitaMyyty(), is(0));
    }

    @Test
    public void kassanMyytyjenMaukkaidenLounaidenMaaraAlussaOikea() {
        assertThat(kassa.maukkaitaLounaitaMyyty(), is(0));
    }    
    
//    käteisosto toimii sekä edullisten että maukkaiden lounaiden osalta
//    jos maksu riittävä: kassassa oleva rahamäärä kasvaa lounaan hinnalla ja vaihtorahan suuruus on oikea
    @Test
    public void edullinenLounasOstoKateinen() {
        int vaihtoraha = kassa.syoEdullisesti(500);
        assertThat(vaihtoraha, is(260));
        assertThat(kassa.edullisiaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100240));
    }
    
    @Test
    public void edullinenLounasOstoKateinenLiianVahanRahaa() {
        int vaihtoraha = kassa.syoEdullisesti(200);
        assertThat(vaihtoraha, is(200));
        assertThat(kassa.edullisiaLounaitaMyyty(), is(0));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }    
    
    @Test
    public void edullinenLounasOstoKortti() {
        Maksukortti kortti = new Maksukortti(1000);
        boolean maksu = kassa.syoEdullisesti(kortti);
        assertThat(maksu, is(true));
        assertThat(kortti.saldo(), is(760));
        assertThat(kassa.edullisiaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
    
    @Test
    public void edullinenLounasOstoKorttiLiianVahanRahaa() {
        Maksukortti kortti = new Maksukortti(238);
        boolean maksu = kassa.syoEdullisesti(kortti);
        assertThat(maksu, is(false));
        assertThat(kortti.saldo(), is(238));
        assertThat(kassa.edullisiaLounaitaMyyty(), is(0));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
    
    @Test
    public void edullinenLounasOstoKateinenTasaraha() {
        int vaihtoraha = kassa.syoEdullisesti(240);
        assertThat(vaihtoraha, is(0));
        assertThat(kassa.edullisiaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100240));
    }    
    
    @Test
    public void edullinenLounasOstoKorttiTasaraha() {
        Maksukortti kortti = new Maksukortti(240);
        boolean maksu = kassa.syoEdullisesti(kortti);
        assertThat(maksu, is(true));
        assertThat(kortti.saldo(), is(0));
        assertThat(kassa.edullisiaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
        
    @Test
    public void maukasLounasOstoKateinen() {
        int vaihtoraha = kassa.syoMaukkaasti(500);
        assertThat(vaihtoraha, is(100));
        assertThat(kassa.maukkaitaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100400));
    }
    
    @Test
    public void maukasLounasOstoKateinenLiianVahanRahaa() {
        int vaihtoraha = kassa.syoMaukkaasti(300);
        assertThat(vaihtoraha, is(300));
        assertThat(kassa.maukkaitaLounaitaMyyty(), is(0));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }    
    
    @Test
    public void maukasLounasOstoKortti() {
        Maksukortti kortti = new Maksukortti(1000);
        boolean maksu = kassa.syoMaukkaasti(kortti);
        assertThat(maksu, is(true));
        assertThat(kortti.saldo(), is(600));
        assertThat(kassa.maukkaitaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
    
    @Test
    public void maukasLounasOstoKorttiLiianVahanRahaa() {
        Maksukortti kortti = new Maksukortti(238);
        boolean maksu = kassa.syoMaukkaasti(kortti);
        assertThat(maksu, is(false));
        assertThat(kortti.saldo(), is(238));
        assertThat(kassa.maukkaitaLounaitaMyyty(), is(0));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
    
    @Test
    public void maukasLounasOstoKateinenTasaraha() {
        int vaihtoraha = kassa.syoMaukkaasti(400);
        assertThat(vaihtoraha, is(0));
        assertThat(kassa.maukkaitaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100400));
    }    
    
    @Test
    public void maukasLounasOstoKorttiTasaraha() {
        Maksukortti kortti = new Maksukortti(400);
        boolean maksu = kassa.syoMaukkaasti(kortti);
        assertThat(maksu, is(true));
        assertThat(kortti.saldo(), is(0));
        assertThat(kassa.maukkaitaLounaitaMyyty(), is(1));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
    
//    kortille rahaa ladattaessa kortin saldo muuttuu ja kassassa oleva rahamäärä kasvaa ladatulla summalla
    @Test
    public void lataaRahaaKortille() {
        Maksukortti kortti = new Maksukortti(200);
        kassa.lataaRahaaKortille(kortti, 600);
        assertThat(kortti.saldo(), is(800));
        assertThat(kassa.kassassaRahaa(), is(100600));
    }
    
    @Test
    public void lataaNegatiivinenMaaraRahaaKortille() {
        Maksukortti kortti = new Maksukortti(200);
        kassa.lataaRahaaKortille(kortti, -200);
        assertThat(kortti.saldo(), is(200));
        assertThat(kassa.kassassaRahaa(), is(100000));
    }
}

File: laskarit/viikko2/unicafe/src/tests/kassapaate_test.py
import unittest
from kassapaate import Kassapaate
from maksukortti import Maksukortti

class TestKassapaate(unittest.TestCase):
	def setUp(self):
		self.kassa = Kassapaate()
		self.kortti = Maksukortti(1000)

	def test_alustettaessa_oikea_maara_rahaa(self):
		self.assertEqual(self.kassa.kassassa_rahaa, 100000)

	def test_alustettaessa_oikea_maara_myytyja_edullisia_lounaita(self):
		self.assertEqual(self.kassa.edulliset, 0)

	def test_alustettaessa_oikea_maara_myytyja_maukkaita_lounaita(self):
		self.assertEqual(self.kassa.maukkaat, 0)

	# Käteisosto toimii sekä edullisten että maukkaiden lounaiden osalta
    # Jos maksu riittävä: kassassa oleva rahamäärä kasvaa lounaan hinnalla ja vaihtorahan suuruus on oikea
	def test_kateisosto_edullinen_palauttaa_oikein(self):
		self.assertEqual(self.kassa.syo_edullisesti_kateisella(500), 260)
	
	def test_kateisosto_edullinen_kassassa_rahaa_oikein(self):
		self.kassa.syo_edullisesti_kateisella(500)
		self.assertEqual(self.kassa.kassassa_rahaa, 100240)

	def test_kateisosto_maukkaat_palauttaa_oikein(self):
		self.assertEqual(self.kassa.syo_maukkaasti_kateisella(500), 100)
	
	def test_kateisosto_maukkaat_kassassa_rahaa_oikein(self):
		self.kassa.syo_maukkaasti_kateisella(500)
		self.assertEqual(self.kassa.kassassa_rahaa, 100400)

    # Jos maksu on riittävä: myytyjen lounaiden määrä kasvaa
	def test_kateisosto_edullinen_kassassa_lounaat_oikein(self):
		self.kassa.syo_edullisesti_kateisella(500)
		self.assertEqual(self.kassa.edulliset, 1)

	def test_kateisosto_maukkaat_kassassa_lounaat_oikein(self):
		self.kassa.syo_maukkaasti_kateisella(500)
		self.assertEqual(self.kassa.maukkaat, 1)

    # Jos maksu ei ole riittävä: kassassa oleva rahamäärä ei muutu, kaikki rahat palautetaan vaihtorahana ja myytyjen lounaiden määrässä ei muutosta
	def test_kateisosto_edullinen_kassa_palauttaa_oikein_liian_pienelle_summalle(self):
		self.assertEqual(self.kassa.syo_edullisesti_kateisella(200), 200)

	def test_kateisosto_edullinen_kassan_rahat_ei_muutu(self):
		self.kassa.syo_edullisesti_kateisella(200)
		self.assertEqual(self.kassa.kassassa_rahaa, 100000)

	def test_kateisosto_edullinen_kassa_ei_lisaa_myytyja_lounaita_liian_pienelle_summalle(self):
		self.kassa.syo_edullisesti_kateisella(200)
		self.assertEqual(self.kassa.edulliset, 0)

	def test_kateisosto_maukkaat_kassa_palauttaa_oikein_liian_pienelle_summalle(self):
		self.assertEqual(self.kassa.syo_maukkaasti_kateisella(300), 300)

	def test_kateisosto_maukkaat_kassan_rahat_ei_muutu(self):
		self.kassa.syo_maukkaasti_kateisella(300)
		self.assertEqual(self.kassa.kassassa_rahaa, 100000)

	def test_kateisosto_maukkaat_kassa_ei_lisaa_myytyja_lounaita_liian_pienelle_summalle(self):
		self.kassa.syo_maukkaasti_kateisella(300)
		self.assertEqual(self.kassa.maukkaat, 0)

	# Korttiosto toimii sekä edullisten että maukkaiden lounaiden osalta
    # Jos kortilla on tarpeeksi rahaa, veloitetaan summa kortilta ja palautetaan True
	def test_korttiosto_edullinen_veloittaa_kortilta_oikein(self):
		self.kassa.syo_edullisesti_kortilla(self.kortti)
		self.assertEqual(str(self.kortti), 'saldo: 7.6')

	def test_korttiosto_edullinen_palauttaa_true(self):
		self.assertEqual(self.kassa.syo_edullisesti_kortilla(self.kortti), True)

	def test_korttiosto_maukkaat_veloittaa_kortilta_oikein(self):
		self.kassa.syo_maukkaasti_kortilla(self.kortti)
		self.assertEqual(str(self.kortti), 'saldo: 6.0')

	def test_korttiosto_maukkaat_palauttaa_true(self):
		self.assertEqual(self.kassa.syo_maukkaasti_kortilla(self.kortti), True)

    # Jos kortilla on tarpeeksi rahaa, myytyjen lounaiden määrä kasvaa
	def test_korttiosto_edullinen_myytyjen_lounaiden_maara_kasvaa(self):
		self.kassa.syo_edullisesti_kortilla(self.kortti)
		self.assertEqual(self.kassa.edulliset, 1)

	def test_korttiosto_maukkaat_myytyjen_lounaiden_maara_kasvaa(self):
		self.kassa.syo_maukkaasti_kortilla(self.kortti)
		self.assertEqual(self.kassa.maukkaat, 1)

    # Jos kortilla ei ole tarpeeksi rahaa, kortin rahamäärä ei muutu, myytyjen lounaiden määrä muuttumaton ja palautetaan False
	def test_korttiosto_edullinen_ei_veloita_kortilta_jos_rahaa_liian_vahan(self):
		koyha = Maksukortti(100)
		self.kassa.syo_edullisesti_kortilla(koyha)
		self.assertEqual(str(koyha), 'saldo: 1.0')

	def test_korttiosto_edullinen_palauttaa_false(self):
		koyha = Maksukortti(100)
		self.assertEqual(self.kassa.syo_edullisesti_kortilla(koyha), False)

	def test_korttiosto_edullinen_ei_kasvata_myytyjen_lounaiden_maaraa_jos_rahaa_liian_vahan(self):
		koyha = Maksukortti(100)
		self.kassa.syo_edullisesti_kortilla(koyha)
		self.assertEqual(self.kassa.edulliset, 0)

	def test_korttiosto_maukkaat_ei_veloita_kortilta_jos_rahaa_liian_vahan(self):
		koyha = Maksukortti(100)
		self.kassa.syo_maukkaasti_kortilla(koyha)
		self.assertEqual(str(koyha), 'saldo: 1.0')

	def test_korttiosto_maukkaat_palauttaa_false(self):
		koyha = Maksukortti(100)
		self.assertEqual(self.kassa.syo_maukkaasti_kortilla(koyha), False)

	def test_korttiosto_maukkaat_ei_kasvata_myytyjen_lounaiden_maaraa_jos_rahaa_liian_vahan(self):
		koyha = Maksukortti(100)
		self.kassa.syo_maukkaasti_kortilla(koyha)
		self.assertEqual(self.kassa.maukkaat, 0)

    # Kassassa oleva rahamäärä ei muutu kortilla ostettaessa
	def test_korttiosto_edullinen_ei_kasvata_kassan_rahamaaraa(self):
		self.kassa.syo_edullisesti_kortilla(self.kortti)
		self.assertEqual(self.kassa.kassassa_rahaa, 100000)

	def test_korttiosto_maukkaat_ei_kasvata_kassan_rahamaaraa(self):
		self.kassa.syo_maukkaasti_kortilla(self.kortti)
		self.assertEqual(self.kassa.kassassa_rahaa, 100000)

	# Kortille rahaa ladattaessa kortin saldo muuttuu ja kassassa oleva rahamäärä kasvaa ladatulla summalla
	def test_kortin_lataus_kasvattaa_kassaa(self):
		self.kassa.lataa_rahaa_kortille(self.kortti, 1000)
		self.assertEqual(self.kassa.kassassa_rahaa, 101000)

	def test_kortin_negatiivinen_lataus_ei_kasvata_kassaa(self):
		self.kassa.lataa_rahaa_kortille(self.kortti, -500)
		self.assertEqual(self.kassa.kassassa_rahaa, 100000)

	# Kortin toiminta parempi testata ainoastaan maksukortti_test.py:ssä ?
	def test_kortin_lataus_kasvattaa_kortin_saldoa(self):
		self.kassa.lataa_rahaa_kortille(self.kortti, 1000)
		self.assertEqual(str(self.kortti), 'saldo: 20.0')

	def test_kortin_negatiivinen_lataus_ei_kasvata_kortin_saldoa(self):
		self.kassa.lataa_rahaa_kortille(self.kortti, -500)
		self.assertEqual(str(self.kortti), 'saldo: 10.0')
